book-content/chapters/08-classes.md

```diff
@@ -0,0 +1,1044 @@
+  const canvasNode = new CanvasNode();
+
+  expect(canvasNode.x).toEqual(0);
+  expect(canvasNode.y).toEqual(0);
+
+  canvasNode.move(10, 20);
+
+  expect(canvasNode.x).toEqual(10);
+  expect(canvasNode.y).toEqual(20);
+});
+```
+
+Currently, there is an error under the `move` method call because the `CanvasNode` class does not have a `move` method.
+
+Your task is to add a `move` method to the `CanvasNode` class that will update the `x` and `y` properties to the new location.
+
+<Exercise title="Exercise 2: Implementing Class Methods" filePath="/src/030-classes/109-class-methods.problem.ts"></Exercise>
+
+### Exercise 3: Implement a Getter
+
+Let's continue working with the `CanvasNode` class, which now has a constructor that accepts an optional argument, renamed to `position`. This `position` is an object that replaces the individual `x` and `y` we had before:
+
+```typescript
+class CanvasNode {
+  x: number;
+  y: number;
+
+  constructor(position?: { x: number; y: number }) {
+    this.x = position?.x ?? 0;
+    this.y = position?.y ?? 0;
+  }
+
+  move(x: number, y: number) {
+    this.x = x;
+    this.y = y;
+  }
+}
+```
+
+In these test cases, there are errors accessing the `position` property since it is not currently a property of the `CanvasNode` class:
+
+```ts twoslash
+// @errors: 2339
+import { it, expect } from "vitest";
+
+class CanvasNode {
+  x: number;
+  y: number;
+
+  constructor(position?: { x: number; y: number }) {
+    this.x = position?.x ?? 0;
+    this.y = position?.y ?? 0;
+  }
+
+  move(x: number, y: number) {
+    this.x = x;
+    this.y = y;
+  }
+}
+
+// ---cut---
+it("Should be able to move", () => {
+  const canvasNode = new CanvasNode();
+
+  expect(canvasNode.position).toEqual({ x: 0, y: 0 });
+
+  canvasNode.move(10, 20);
+
+  expect(canvasNode.position).toEqual({ x: 10, y: 20 });
+});
+
+it("Should be able to receive an initial position", () => {
+  const canvasNode = new CanvasNode({
+    x: 10,
+    y: 20,
+  });
+
+  expect(canvasNode.position).toEqual({ x: 10, y: 20 });
+});
+```
+
+Your task is to update the `CanvasNode` class to include a `position` getter that will allow for the test cases to pass.
+
+<Exercise title="Exercise 3: Implement a Getter" filePath="/src/030-classes/111-getters.problem.ts"></Exercise>
+
+### Exercise 4: Implement a Setter
+
+The `CanvasNode` class has been updated so that `x` and `y` are now private properties:
+
+```typescript
+class CanvasNode {
+  #x: number;
+  #y: number;
+
+  constructor(position?: { x: number; y: number }) {
+    this.#x = position?.x ?? 0;
+    this.#y = position?.y ?? 0;
+  }
+
+  // your `position` getter method here
+
+  // move method as before
+}
+```
+
+The `#` in front of the `x` and `y` properties means they are `readonly` and can't be modified directly outside of the class. In addition, when a getter is present without a setter, its property will also be treated as `readonly`, as seen in this test case:
```

Comment:
```suggestion
The `#` in front of the `x` and `y` properties means they are private and can't be modified directly outside of the class. In addition, when a getter is present without a setter, its property will also be treated as `readonly`, as seen in this test case:
```

In general, something weird is going on here and in subsequent paragraphs. You mention `readonly`, and getters without setters but there are no getters here. There is a mismatch between text and the presented code.

Reply:
I know that getter was supposed to be introduced by the reader in the previous exercise but it's still confusing to me. You didn't even introduce getters and setters at all before, and now you softly expect the reader to be aware of them.

Those exercises are about classes but now you present a simple object type with a simple property as a validation for this statement:
> In addition, when a getter is present without a setter, its property will also be treated as `readonly`, as seen in this test case:

I don't feel like the example is related to this paragraph. That example has a readonly property because it has a readonly modifier on it - not because it's a getter-only property.